### PR TITLE
Batch NDCG

### PR DESCRIPTION
**Pull request by donglihe-hub**
## What does this PR do?

The original NDCG metric calculates scores for one instance at a time, which is inefficient. The new NDCG metric calculate scores in batch.

### Performance Test Settings:
Number of labels = 100

Batch size = 40
Number of batchs = 100
Effective number of validation samples = 4000

### Results:
1c07d1925b9145c64f09d44690f1112

## Test CLI & API (`bash tests/autotest.sh`)
Test APIs used by main.py.
- [ ] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
If any new APIs are added, please check if the description of the APIs is added to API document. 
- [ ] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_changed_document.sh`)
If any APIs in quickstarts or tutorials are modified, please run this test to check if the current examples can run correctly after the modified APIs are released.

### CODE REVIEW

File: libmultilabel/nn/metrics.py [hunk 1]
```diff
@@ -60,18 +62,37 @@ class NDCG(Metric):
...
-
-    def _metric(self, preds, target):
-        return retrieval_normalized_dcg(preds, target, k=self.top_k)
+        score = self.score / self.num_sample
+        return score
+
+    def _dcg(self, preds, target):
+        _, sorted_top_k_idx = torch.topk(preds, k=self.top_k)
+        gains = target.take_along_dim(sorted_top_k_idx, 1)
+        dcg = (gains * self.discount).sum(1)
```
Comment (Sinacam) [inline libmultilabel/nn/metrics.py:84]: Please use `torch.dot`.

**Review by Sinacam — COMMENTED** [review]
We should review the use of plurals vs singulars.

File: libmultilabel/nn/metrics.py (hunk 1 above)
Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py]: Please refer to https://discuss.pytorch.org/t/dot-product-batch-wise/9746/11. (a * b).sum(dim=1) has the best performance regarding batch dot product.

File: libmultilabel/nn/metrics.py [hunk 2]
```diff
@@ -60,18 +62,37 @@ class NDCG(Metric):
...
+        self.add_state("num_sample", default=torch.tensor(0, dtype=torch.int64), dist_reduce_fx="sum")
+        self.add_state("discount", default=1.0 / torch.log2(torch.arange(self.top_k) + 2.0))
 
     def update(self, preds, target):
         assert preds.shape == target.shape
-        # implement batch-wise calculations instead of storing results of all batches
-        self.ndcg += [self._metric(p, t) for p, t in zip(preds, target)]
+        dcg = self._dcg(preds, target)
+        idcg = self._idcg(target)
+        ndcg = torch.nan_to_num(dcg / idcg, nan=0.0)
```
[the remark this answers is not in the document]
Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py]: Add comments about hacking to the idcg.

File: libmultilabel/nn/metrics.py [hunk 3]
```diff
@@ -60,18 +62,37 @@ class NDCG(Metric):
...
+        return score
+
+    def _dcg(self, preds, target):
+        _, sorted_top_k_idx = torch.topk(preds, k=self.top_k)
+        gains = target.take_along_dim(sorted_top_k_idx, 1)
+        dcg = (gains * self.discount).sum(1)
+        return dcg
+
+    def _idcg(self, target):
+        """optimized idcg for multilabel classification"""
```
[the remark this answers is not in the document]
Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py]: Removed the comment. My original purpose is to tell readers idcg here requires "target" to be binary.

File: libmultilabel/nn/metrics.py [hunk 4]
```diff
@@ -60,18 +62,37 @@ class NDCG(Metric):
...
+        ndcg = torch.nan_to_num(dcg / idcg, nan=0.0)
+        self.score += ndcg.sum()
+        self.num_sample += preds.shape[0]
 
     def compute(self):
-        return torch.stack(self.ndcg).mean()
-
-    def _metric(self, preds, target):
-        return retrieval_normalized_dcg(preds, target, k=self.top_k)
+        score = self.score / self.num_sample
```
[the remark this answers is not in the document]
Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py]: Done

**Comment by donglihe-hub** [comment]
> We should review the use of plurals vs singulars.

A question that has haunted me for years is that why it is preds and target, rather than preds and targets or pred and target. Therefore, I prefer using singular form for all variable except for those concepts that have been accepted by the ML community.

File: libmultilabel/nn/metrics.py [hunk 5]
```diff
@@ -60,18 +62,33 @@ class NDCG(Metric):
...
+        _, sorted_top_k_idx = torch.topk(preds, k=self.top_k)
+        gains = target.take_along_dim(sorted_top_k_idx, dim=1)
+        # best practice for batch dot product: https://discuss.pytorch.org/t/dot-product-batch-wise/9746/11
+        return (gains * discount).sum(dim=1)
 
-    def _metric(self, preds, target):
-        return retrieval_normalized_dcg(preds, target, k=self.top_k)
+    def _idcg(self, target, discount):
+        cum_discount = discount.cumsum(dim=0)
+        idx = target.sum(dim=1).clamp(max=self.top_k) - 1
```
Comment (Sinacam) [inline libmultilabel/nn/metrics.py:90]: `idx = target.sum(dim=1).clamp(min=0, max=self.top_k - 1)`

Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py]: test drafted text

File: libmultilabel/nn/metrics.py [hunk 6]
```diff
@@ -45,6 +45,7 @@ class NDCG(Metric):
     https://scikit-learn.org/stable/modules/generated/sklearn.metrics.ndcg_score.html
     Please find the formal definition here:
     https://nlp.stanford.edu/IR-book/html/htmledition/evaluation-of-ranked-retrieval-results-1.html
+    The target has to be a binary tensor.
```
Comment (cjlin1) [inline libmultilabel/nn/metrics.py:48]: add

We do not use NDCG in ?? because of ??

Comment (donglihe-hub) [reply libmultilabel/nn/metrics.py:48]: I'm not pretty sure what should be filled in the ??. I might just delete these problematic illustrations to not confuse anyone.

**Review by cjlin1 — APPROVED** [review]
